larq_compute_engine/mlir/python/end2end_verify.cc

```diff
@@ -0,0 +1,132 @@
+    std::cerr << "Error at line " << __LINE__ << std::endl; \
+    exit(1);                                                \
+  }
+
+std::vector<float> parseFile(const char* filename) {
+  std::ifstream file(filename);
+  if (!file) {
+    std::cerr << "Could not open file: " << filename << std::endl;
+    exit(1);
+  }
+  std::vector<float> result;
+  for (;;) {
+    float x;
+    file >> x;
+    if (!file) break;
+    result.push_back(x);
+  }
+  return result;
+}
+
+int main(int argc, char* argv[]) {
+  if (argc != 4) {
+    std::cerr << "end2end_verify <tflite model> <input txt> <output txt>"
+              << std::endl;
+    return 1;
+  }
+  const char* filename_model = argv[1];
+  const char* filename_input = argv[2];
+  const char* filename_output = argv[3];
+
+  std::vector<float> input = parseFile(filename_input);
+  std::vector<float> output = parseFile(filename_output);
+
+  // Load model
+  std::unique_ptr<tflite::FlatBufferModel> model =
+      tflite::FlatBufferModel::BuildFromFile(filename_model);
+  TFLITE_MINIMAL_CHECK(model != nullptr);
+
+  // Build the interpreter
+  tflite::ops::builtin::BuiltinOpResolver resolver;
+  compute_engine::tflite::RegisterLCECustomOps(&resolver);
+
+  InterpreterBuilder builder(*model, resolver);
+  std::unique_ptr<Interpreter> interpreter;
+  builder(&interpreter);
+  TFLITE_MINIMAL_CHECK(interpreter != nullptr);
+
+  // Allocate tensor buffers.
+  TFLITE_MINIMAL_CHECK(interpreter->AllocateTensors() == kTfLiteOk);
+
+  auto input_ids = interpreter->inputs();
+  auto output_ids = interpreter->outputs();
+
+  TFLITE_MINIMAL_CHECK(input_ids.size() == 1);
+  TFLITE_MINIMAL_CHECK(output_ids.size() == 1);
+  int input_id = input_ids[0];
+  int output_id = output_ids[0];
+
+  const auto* input_tensor = interpreter->tensor(input_id);
+  float* input_tensor_data = interpreter->typed_tensor<float>(input_id);
+  TFLITE_MINIMAL_CHECK(input_tensor_data != nullptr);
+  TFLITE_MINIMAL_CHECK(input_tensor->dims != nullptr);
+  TFLITE_MINIMAL_CHECK(input_tensor->dims->size != 0);
+
+  int num_input_elements = 1;
+  for (int i = 0; i < input_tensor->dims->size; ++i) {
+    num_input_elements *= input_tensor->dims->data[i];
+  }
+  TFLITE_MINIMAL_CHECK(num_input_elements == (int)input.size());
+
+  const auto* output_tensor = interpreter->tensor(output_id);
+  const float* output_tensor_data = interpreter->typed_tensor<float>(output_id);
+  TFLITE_MINIMAL_CHECK(output_tensor_data != nullptr);
+  TFLITE_MINIMAL_CHECK(output_tensor->dims != nullptr);
+  TFLITE_MINIMAL_CHECK(output_tensor->dims->size == 2);
+  TFLITE_MINIMAL_CHECK(output_tensor->dims->data[0] == 1);
+  int num_output_elements = output_tensor->dims->data[1];
+  TFLITE_MINIMAL_CHECK(num_output_elements == (int)output.size());
+
+  // Do inference twice, to check if the warmpup run and internal temporary
+  // buffers work as expected
+  for (int run = 0; run < 2; ++run) {
+    // Fill input buffers
```

Comment:
this can be moved outside the for loop. you don't need to fill the input buffer with the same values twice.

Reply:
That depends if the input buffer is used to store intermediate activations, i.e., tflite may share the memory for different tensors.

Reply:
I highly doubt that the tflite invalidates the input buffer (e.g the case the input buffer comes from another part of the code such as image pre-processing). but anyway this is not critical and we keep it the same way as it is.